pkg/services/query/errors.go

```diff
@@ -1,17 +1,19 @@
 package query
 
-import "fmt"
+import (
+	"errors"
 
-// ErrBadQuery returned whenever request is malformed and must contain a message
-// suitable to return in API response.
-type ErrBadQuery struct {
-	Message string
-}
+	"github.com/grafana/grafana/pkg/util"
+	"github.com/grafana/grafana/pkg/util/errutil"
+)
 
-func NewErrBadQuery(msg string) *ErrBadQuery {
-	return &ErrBadQuery{Message: msg}
-}
+var ErrQueryValidationFailure = errutil.NewBase(errutil.StatusBadRequest, "query.validationError").MustTemplate("query validation failed: {{ .Error }}", errutil.WithPublic("{{ .Public.Message }}"))
```

Comment:
@sakjur @wbrowne after some considerations I wonder if we're using this templated error correctly/as it was thought. We basically re-uses this error base for 4 different query validation errors - it will have the same message id, same status, but different underlying error and public message. 

If we're planning to use the message id with the purpose of client creating a public message from it the only thing it could do is basically say "Query validation error".

Wonder if the correct way rather is to define 4 different base errors for each one of these query validation errors.

Thoughts?

Reply:
Try to think in terms of "units of translation". If you can divide this into four different query validation errors that will allow you to provide useful context when translating a message, that'd be preferable.

If `.Public.Message` in this case is unpredictable and depends on the data source, we're unlikely to translate that, so it should be ok to limit the translatable part to "Query validation error", which would make the public message something like "Query validation error: {{ .Public.Message }}" which can be translated into 🇸🇪"Frågevalideringsfel: {{ .Public.Message }}" by the frontend.

Reply:
Still confused about the purpose of message id and translation I think. Are we suppose to provide the public message template to the frontend together with a public payload and frontend will use these to translate? Or do you mean that the frontend will receive the public message in english and try to translate parts of it that it understands?

Reply:
Ohh I see now that when building the error from template the PublicPayload is populated https://github.com/grafana/grafana/blob/bd19a06fa951e140e939e0a92bb49b70f4b952f1/pkg/util/errutil/template.go#L113

That to me sounds like we should have 4 different errors instead of 1 so that we can return proper public payload, e.g. 
for https://github.com/grafana/grafana/blob/b7cf1cfcb9aa16b5897530a0a265e1e2f7692bea/pkg/services/query/query.go#L120

the public payload should then include the `refID` if I understand this correctly.

Reply:
Yes, exactly like that 🙂